lib/Dialect/CGGI/Transforms/BooleanLineVectorizer.cpp

```diff
@@ -0,0 +1,224 @@
+#include "lib/Dialect/CGGI/Transforms/BooleanLineVectorizer.h"
+
+#include <string>
+
+#include "lib/Dialect/CGGI/IR/CGGIAttributes.h"
+#include "lib/Dialect/CGGI/IR/CGGIOps.h"
+#include "lib/Dialect/LWE/IR/LWEAttributes.h"
+#include "lib/Graph/Graph.h"
+#include "llvm/include/llvm/ADT/TypeSwitch.h"          // from @llvm-project
+#include "llvm/include/llvm/Support/Debug.h"           // from @llvm-project
+#include "mlir/include/mlir/Analysis/SliceAnalysis.h"  // from @llvm-project
+#include "mlir/include/mlir/Analysis/TopologicalSortUtils.h"  // from @llvm-project
+#include "mlir/include/mlir/Dialect/Arith/IR/Arith.h"    // from @llvm-project
+#include "mlir/include/mlir/Dialect/Tensor/IR/Tensor.h"  // from @llvm-project
+#include "mlir/include/mlir/IR/Dialect.h"                // from @llvm-project
+#include "mlir/include/mlir/IR/Visitors.h"               // from @llvm-project
+#include "mlir/include/mlir/Support/LLVM.h"              // from @llvm-project
+#include "mlir/include/mlir/Transforms/GreedyPatternRewriteDriver.h"  // from @llvm-project
+#include "mlir/include/mlir/Transforms/Passes.h"  // from @llvm-project
+
+#define DEBUG_TYPE "bool-line-vectorizer"
+
+namespace mlir {
+namespace heir {
+namespace cggi {
+
+#define GEN_PASS_DEF_BOOLEANLINEVECTORIZER
+#include "lib/Dialect/CGGI/Transforms/Passes.h.inc"
+
+bool areCompatibleBool(Operation *lhs, Operation *rhs) {
+  if (
+      // lhs->getName() != rhs->getName() ||
+      lhs->getDialect() != rhs->getDialect() ||
+      lhs->getResultTypes() != rhs->getResultTypes() ||
+      lhs->getAttrs() != rhs->getAttrs()) {
+    return false;
+  }
+  return OpTrait::hasElementwiseMappableTraits(lhs);
```

Comment:
this might be somewhat fragile right? since there might be other elementwise operations that aren't boolean gates..? maybe add a comment or a TODO if this can be made better with a BooleanPackableGate trait on the CGGI_BinaryGateOp's?